extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergManualRefreshTableLocationProvider.java

```diff
@@ -0,0 +1,124 @@
+    // AbstractTableLocationProvider implementation
+    // ------------------------------------------------------------------------------------------------------------------
+
+    @Override
+    public String getImplementationName() {
+        return IMPLEMENTATION_NAME;
+    }
+
+    @Override
+    public synchronized void refresh() {
+        // There should be no refresh service for this provider.
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public synchronized void update() {
+        update(adapter.getCurrentSnapshot(tableIdentifier));
+    }
+
+    @Override
+    public synchronized void update(final long snapshotId) {
+        final List<Snapshot> snapshots = adapter.listSnapshots(tableIdentifier);
+
+        final Snapshot snapshot = snapshots.stream()
+                .filter(s -> s.snapshotId() == snapshotId).findFirst()
+                .orElse(null);
+
+        if (snapshot == null) {
+            throw new IllegalArgumentException(
+                    "Snapshot " + snapshotId + " was not found in the list of snapshots for table " + tableIdentifier
+                            + ". Snapshots: " + snapshots);
+        }
+        update(snapshot);
+    }
+
+    @Override
+    public synchronized void update(final Snapshot snapshot) {
+        // Verify that the input snapshot is newer (higher in sequence number) than the current snapshot.
+        if (snapshot.sequenceNumber() <= locationKeyFinder.snapshot.sequenceNumber()) {
+            throw new IllegalArgumentException(
+                    "Update snapshot sequence number (" + snapshot.sequenceNumber()
+                            + ") must be higher than the current snapshot sequence number ("
+                            + locationKeyFinder.snapshot.sequenceNumber() + ") for table " + tableIdentifier);
+        }
+        // Update the snapshot.
+        locationKeyFinder.snapshot = snapshot;
+        refreshSnapshot();
+    }
+
+    /**
+     * Refresh the table location provider with the latest snapshot from the catalog. This method will identify new
+     * locations and removed locations.
+     */
+    private void refreshSnapshot() {
+        beginTransaction(this);
+        final Set<ImmutableTableLocationKey> missedKeys = new HashSet<>(getTableLocationKeys());
+        locationKeyFinder.findKeys(tlk -> {
+            missedKeys.remove(tlk);
+            handleTableLocationKeyAdded(tlk, this);
+        });
+        missedKeys.forEach(tlk -> handleTableLocationKeyRemoved(tlk, this));
+        endTransaction(this);
+        setInitialized();
+    }
+
+    // ------------------------------------------------------------------------------------------------------------------
+    // SubscriptionAggregator implementation
+    // ------------------------------------------------------------------------------------------------------------------
+
+    @Override
+    protected void activateUnderlyingDataSource() {
+        refreshSnapshot();
```

Comment:
This probably needs to be guarded by some initialization filter.
We don't want to `refreshSnapshot()` for the same snapshot more than once.